src/react/components/InterfaceEditor.tsx

```diff
@@ -0,0 +1,945 @@
+/*
+   This file is part of Astarte.
+
+   Copyright 2020 Ispirata Srl
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import React, { useCallback, useEffect, useState } from 'react';
+import {
+  Accordion,
+  Badge,
+  Button,
+  Card,
+  Col,
+  Container,
+  Form,
+  InputGroup,
+  Modal,
+  Row,
+} from 'react-bootstrap';
+import { AstarteInterface, AstarteMapping } from 'astarte-client';
+import _ from 'lodash';
+
+import MappingEditor from './MappingEditor';
+
+interface FormControlWarningProps {
+  message?: string;
+}
+
+const FormControlWarning = ({ message }: FormControlWarningProps): React.ReactElement | null => {
+  if (!message) {
+    return null;
+  }
+  return <div className="warning-feedback">{message}</div>;
+};
+
+interface MappingRowProps {
+  className?: string;
+  mapping: AstarteMapping;
+  onEdit: () => void;
+  onDelete: () => void;
+}
+
+const reliabilityToLabel = {
+  unreliable: 'Unreliable',
+  guaranteed: 'Guaranteed',
+  unique: 'Unique',
+};
+
+const retentionToLabel = {
+  discard: 'Discard',
+  volatile: 'Volatile',
+  stored: 'Stored',
+};
+
+const databaseRetentionToLabel = {
+  no_ttl: 'No TTL',
+  use_ttl: 'Use TTL',
+};
+
+const MappingRow = ({ className, mapping, onEdit, onDelete }: MappingRowProps) => {
+  return (
+    <Accordion>
+      <Card className={className}>
+        <Accordion.Toggle
+          eventKey={mapping.endpoint}
+          as={Card.Header}
+          className="d-flex align-items-center flex-wrap"
+        >
+          <span className="flex-grow-1">
+            <Badge variant="secondary">{mapping.type}</Badge>
+            <Button className="text-left text-truncate" variant="link">
+              {mapping.endpoint}
+            </Button>
+          </span>
+          <Button className="mr-2" variant="outline-primary" onClick={onEdit}>
+            Edit...
+          </Button>
+          <Button variant="outline-primary" onClick={onDelete}>
+            Remove
+          </Button>
+        </Accordion.Toggle>
+        <Accordion.Collapse eventKey={mapping.endpoint}>
+          <Card.Body>
+            {mapping.description && (
```

Comment:
this h5 - p is repeated a lot. i suggest creating both a custom component and an array of label-value tuples.
You can then _map_ the array with the component you just created.
If you want to make it even cooler and easier to understand, filter undefined/default values before mapping them

Reply:
I tried abstracting a custom component as you suggested. However I have to say it really impaired readability.
I personally favor readability over (modest) duplication, so I would keep the component straightforward as it is.
As for the idea of filtering out undefined values, I did already see what was done on the Elm side but I would favor the current approach for the same reason as above.